File: game_server/test/sc/server/helpers/ConverterTest.java
package sc.server.helpers;

import org.junit.Assert;
import org.junit.Ignore;
import org.junit.Test;

import com.thoughtworks.xstream.XStream;

import sc.framework.plugins.IPerspectiveAware;
import sc.framework.plugins.IPerspectiveProvider;
import sc.server.Configuration;

public class ConverterTest
{
	public static class HasSecrets implements IPerspectiveProvider,
			IPerspectiveAware
	{
		public static final Object	hacker = new Object();

		public static final Object	goodFriend = new Object();

		private String				secret = "i-am-secret";

		private String				unimportant = "i-am-unimportant";

		private Object				perspective = null;

		@Override
		public Object getPerspective()
		{
			return this.perspective;
		}

		public void setPerspective(Object o)
		{
			this.perspective = o;
		}

		@Override
		public boolean isVisibleFor(Object viewer, String field)
		{
			return !(field.equals("secret") && hacker.equals(viewer));
		}
	}

	@Test
	public void shouldSerializeSensitiveDataForAuthorizedPeople()
	{
		HasSecrets data = new HasSecrets();
		data.setPerspective(HasSecrets.goodFriend);

		XStream xStream = Configuration.getXStream();
		String msg = xStream.toXML(data);

		Assert.assertNotSame(-1, msg.indexOf(data.secret));
		Assert.assertNotSame(-1, msg.indexOf(data.unimportant));
	}

	@Test
	public void shouldSerializeSensitiveDataForObservers()
	{
		HasSecrets data = new HasSecrets();
		data.setPerspective(null);

		XStream xStream = Configuration.getXStream();
		String msg = xStream.toXML(data);

		Assert.assertNotSame(-1, msg.indexOf(data.secret));
		Assert.assertNotSame(-1, msg.indexOf(data.unimportant));
	}

}

File: server/test/sc/server/helpers/ConverterTest.kt
package sc.server.helpers

import org.junit.Assert
import org.junit.Test
import sc.framework.plugins.IPerspectiveAware
import sc.framework.plugins.IPerspectiveProvider
import sc.helpers.xStream

class ConverterTest {
    class HasSecrets(private val perspective: Any?) : IPerspectiveProvider, IPerspectiveAware {
        val secret = "i-am-secret"
        val unimportant = "i-am-unimportant"
        override fun getPerspective(): Any? {
            return perspective
        }

        override fun isVisibleFor(viewer: Any, field: String): Boolean {
            return !(field == "secret" && hacker == viewer)
        }

        companion object {
            val hacker = Any()
            val goodFriend = Any()
        }
    }

    @Test
    fun shouldSerializeSensitiveDataForAuthorizedPeople() {
        val data = HasSecrets(HasSecrets.goodFriend)
        val msg = xStream.toXML(data)
        Assert.assertNotSame(-1, msg.indexOf(data.secret))
        Assert.assertNotSame(-1, msg.indexOf(data.unimportant))
    }

    @Test
    fun shouldSerializeSensitiveDataForObservers() {
        val data = HasSecrets(null)
        val msg = xStream.toXML(data)
        Assert.assertNotSame(-1, msg.indexOf(data.secret))
        Assert.assertNotSame(-1, msg.indexOf(data.unimportant))
    }
}